File: Core/src/main/java/me/mrkirby153/kcutils/scoreboard/KirbyScoreboard.java
package me.mrkirby153.kcutils.scoreboard;

import me.mrkirby153.kcutils.scoreboard.items.ElementText;
import net.md_5.bungee.api.ChatColor;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.scoreboard.DisplaySlot;
import org.bukkit.scoreboard.Objective;
import org.bukkit.scoreboard.Scoreboard;
import org.bukkit.scoreboard.Team;

import java.util.*;
import java.util.stream.Collectors;

/**
 * Wrapper class for quickly creating scoreboards
 */
public class KirbyScoreboard {

    private final Random r = new Random();
    private Scoreboard scoreboard;
    private Objective sideObjective;
    private ArrayList<ScoreboardElement> scoreboardElements = new ArrayList<>();
    private Set<ScoreboardTeam> teams = new HashSet<>();
    private String[] current = new String[15];

    private boolean debug = false;

    public KirbyScoreboard(String displayName) {
        scoreboard = Bukkit.getScoreboardManager().getNewScoreboard();
        sideObjective = addObjective(displayName, DisplaySlot.SIDEBAR);
    }

    /**
     * Adds a scoreboard element
     *
     * @param element The scoreboard element to add
     */
    public void add(ScoreboardElement element) {
        scoreboardElements.add(element);
    }

    /**
     * Adds a scoreboard element
     *
     * @param text The text to add
     */
    public void add(String text) {
        this.add(new ElementText(text));
    }

    /**
     * Adds a team to the scoreboard
     *
     * @param team The team to add
     */
    public void addTeam(ScoreboardTeam team) {
        teams.add(team);
    }

    /**
     * Renders the scoreboard
     */
    public final void draw() {
        if (debug) {
            System.out.println("--------------");
            System.out.println("Calling preDraw()");
        }
        preDraw();
        if (debug)
            System.out.println("Calling updateTeams()");
        updateTeams();

        if (debug)
            System.out.println("Updating scoreboard elements");
        ArrayList<String> newLines = new ArrayList<>();
        for (ScoreboardElement item : this.scoreboardElements) {
            for (String line : item.getLines()) {
                while (true) {
                    boolean matched = false;
                    for (String otherLines : newLines) {
                        if (line.equals(otherLines)) {
                            line += ChatColor.RESET;
                            matched = true;
                        }
                    }
                    if (!matched)
                        break;
                }
                newLines.add(line);
            }
        }

        HashSet<Integer> toAdd = new HashSet<>();
        HashSet<Integer> toRemove = new HashSet<>();
        for (int i = 0; i < 15; i++) {
            if (i >= newLines.size()) {
                if (current[i] != null) {
                    toRemove.add(i);
                }
                continue;
            }

            if (current[i] == null || !current[i].equals(newLines.get(i))) {
                toRemove.add(i);
                toAdd.add(i);
            }
        }
        for (int i : toRemove) {
            if (current[i] != null) {
                resetScore(current[i]);
                current[i] = null;
            }
        }
        for (int i : toAdd) {
            String newLine = newLines.get(i);
            sideObjective.getScore(newLine).setScore(15 - i);
            current[i] = newLine;
        }

        if (debug)
            System.out.println("Calling postDraw()");
        postDraw();
    }

    /**
     * Enable debug output
     */
    public void enableDebug() {
        debug = true;
    }

    /**
     * Gets the scoreboard
     *
     * @return The scoreboard
     */
    public Scoreboard getBoard() {
        return this.scoreboard;
    }

    /**
     * Removes a team from the scoreboard
     *
     * @param team The team to remove
     */
    public void remove(ScoreboardTeam team) {
        teams.remove(team);
    }

    /**
     * Resets the scoreboard (Clears all the lines)
     */
    public void reset() {
        scoreboardElements.clear();
    }

    /**
     * Called after the scoreboard has been rendered
     */
    protected void postDraw() {

    }

    /**
     * Called before the scoreboard is rendered
     */
    protected void preDraw() {

    }

    /**
     * Clears the score of a line on the scoreboard (Removes it)
     *
     * @param line The line to remove
     */
    private void resetScore(String line) {
        scoreboard.resetScores(line);
    }

    /**
     * Updates the teams on the scoreboard
     */
    private void updateTeams() {
        teams.forEach(t -> {
            if (debug)
                System.out.println("Updating team " + t.getTeamName());
            Team scoreboardTeam = null;
            if (scoreboard.getTeam(t.getFilteredName()) != null) {
                scoreboardTeam = scoreboard.getTeam(t.getFilteredName());
            }
            if (scoreboardTeam == null)
                scoreboardTeam = scoreboard.registerNewTeam(t.getFilteredName());
            String prefix = t.getColor() + "";
            if (t.showPrefix()) {
                prefix += String.format("[%s]", t.getPrefix().toUpperCase());
            }
            scoreboardTeam.setPrefix(prefix);
            scoreboardTeam.setSuffix(ChatColor.RESET + "");
            scoreboardTeam.setCanSeeFriendlyInvisibles(t.canSeeInvisibles());
            scoreboardTeam.setAllowFriendlyFire(t.isFriendlyFire());

            // Update all the players on the team
            List<String> playersOnTeam = t.getPlayers().stream().map(Bukkit::getPlayer).filter(Objects::nonNull).map(Player::getName).collect(Collectors.toList());
            Set<String> playersOnScoreboardTeam = scoreboardTeam.getEntries();

            List<String> toAdd = playersOnTeam.stream().filter(p -> !playersOnScoreboardTeam.contains(p)).collect(Collectors.toList());
            List<String> toRemove = playersOnScoreboardTeam.stream().filter(p -> !playersOnTeam.contains(p)).collect(Collectors.toList());
            if (debug && toAdd.size() > 0)
                System.out.println("[ADD] " + Arrays.toString(toAdd.toArray(new String[0])));
            if (debug && toRemove.size() > 0)
                System.out.println("[REMOVE] " + Arrays.toString(toRemove.toArray(new String[0])));

            toRemove.forEach(scoreboardTeam::removeEntry);
            toAdd.forEach(scoreboardTeam::addEntry);
        });
        // Remove teams that no longer exist
        scoreboard.getTeams().forEach(team -> {
            boolean exists = false;
            for (ScoreboardTeam t : teams) {
                if (t.getTeamName().startsWith(team.getName())) {
                    exists = true;
                    break;
                }
            }
            if (!exists) {
                if (debug)
                    System.out.println("Unregistering team " + team.getName());
                team.unregister();
            }
        });
    }

    /**
     * Adds an objective to the scoreboard
     *
     * @param displayName The display name of the objective
     * @param displaySlot The slot that the objective is displayed in
     * @param criteria    The criteria of the objective
     * @return The Objective
     */
    protected Objective addObjective(String displayName, DisplaySlot displaySlot, String criteria) {
        Objective o = scoreboard.registerNewObjective("Obj-" + r.nextInt(999999), criteria);
        o.setDisplaySlot(displaySlot);
        o.setDisplayName(displayName);
        return o;
    }

    /**
     * Adds a dummy objective to the scoreboard
     *
     * @param displayName The display name of the objective
     * @param displaySlot The slot the objective is displayed in
     * @return The objective
     */
    protected Objective addObjective(String displayName, DisplaySlot displaySlot) {
        return this.addObjective(displayName, displaySlot, "dummy");
    }

}

File: Bukkit/src/main/kotlin/me/mrkirby153/kcutils/scoreboard/KirbyScoreboard.kt
package me.mrkirby153.kcutils.scoreboard

import me.mrkirby153.kcutils.scoreboard.items.ElementText
import net.kyori.adventure.text.Component
import net.md_5.bungee.api.ChatColor
import org.bukkit.Bukkit
import org.bukkit.scoreboard.Criteria
import org.bukkit.scoreboard.DisplaySlot
import org.bukkit.scoreboard.Objective
import org.bukkit.scoreboard.Scoreboard
import org.bukkit.scoreboard.Team
import java.util.Random

/**
 * Wrapper class for quickly creating scoreboards
 *
 * @param displayName The name of a scoreboard
 */
open class KirbyScoreboard(displayName: Component) {

    private val r = Random()

    constructor(displayName: String) : this(Component.text(displayName))

    /**
     * Gets the scoreboard
     *
     * @return The scoreboard
     */
    val board: Scoreboard = Bukkit.getScoreboardManager().newScoreboard
    private val sideObjective = addObjective(displayName, DisplaySlot.SIDEBAR)
    private val scoreboardElements = ArrayList<ScoreboardElement>()

    /**
     * A list of teams currently registered on the scoreboard
     */
    private val teams = HashSet<ScoreboardTeam>()

    /**
     * The current list of lines in the scoreboard
     */
    private val current = arrayOfNulls<String>(15)

    private var debug = false

    /**
     * Adds a scoreboard element
     *
     * @param element The scoreboard element to add
     */
    fun add(element: ScoreboardElement) {
        scoreboardElements.add(element)
    }

    /**
     * Adds a scoreboard element
     *
     * @param text The text to add
     */
    fun add(text: String) {
        this.add(ElementText(text))
    }

    /**
     * Adds a team to the scoreboard
     *
     * @param team The team to add
     */
    fun addTeam(team: ScoreboardTeam) {
        teams.add(team)
    }

    /**
     * Renders the scoreboard
     */
    fun draw() {
        if (debug) {
            println("--------------")
            println("Calling preDraw()")
        }
        preDraw()
        if (debug)
            println("Calling updateTeams()")
        updateTeams()

        if (debug)
            println("Updating scoreboard elements")
        val newLines = ArrayList<String>()
        for (item in this.scoreboardElements) {
            for (line in item.lines) {
                var l = line
                while (true) {
                    var matched = false
                    for (otherLines in newLines) {
                        if (l == otherLines) {
                            l += ChatColor.RESET
                            matched = true
                        }
                    }
                    if (!matched)
                        break
                }
                newLines.add(l)
            }
        }

        val toAdd = HashSet<Int>()
        val toRemove = HashSet<Int>()
        for (i in 0..14) {
            if (i >= newLines.size) {
                if (current[i] != null) {
                    toRemove.add(i)
                }
                continue
            }

            if (current[i] == null || current[i] != newLines[i]) {
                toRemove.add(i)
                toAdd.add(i)
            }
        }
        for (i in toRemove) {
            if (current[i] != null) {
                resetScore(current[i]!!)
                current[i] = null
            }
        }
        for (i in toAdd) {
            val newLine = newLines[i]
            sideObjective.getScore(newLine).score = 15 - i
            current[i] = newLine
        }

        if (debug)
            println("Calling postDraw()")
        postDraw()
    }

    /**
     * Enable debug output
     */
    fun enableDebug() {
        debug = true
    }

    /**
     * Get the teams registered on the scoreboard
     *
     * @return The teams registered on the scoreboard
     */
    open fun getTeams(): Set<ScoreboardTeam> {
        return teams
    }

    /**
     * Removes a team from the scoreboard
     *
     * @param team The team to remove
     */
    fun remove(team: ScoreboardTeam) {
        teams.remove(team)
    }

    /**
     * Resets the scoreboard (Clears all the lines)
     */
    fun reset() {
        scoreboardElements.clear()
    }

    /**
     * Clears the score of a line on the scoreboard (Removes it)
     *
     * @param line The line to remove
     */
    private fun resetScore(line: String) {
        board.resetScores(line)
    }

    /**
     * Updates the teams on the scoreboard
     */
    private fun updateTeams() {
        getTeams().forEach { t ->
            if (debug)
                println("Updating team " + t.teamName)
            var scoreboardTeam: Team? = null
            if (board.getTeam(t.filteredName) != null) {
                scoreboardTeam = board.getTeam(t.filteredName)
            }
            if (scoreboardTeam == null)
                scoreboardTeam = board.registerNewTeam(t.filteredName)

            val prefix = if (t.showPrefix) t.prefix?.uppercase() ?: "" else ""
            scoreboardTeam.prefix(Component.empty().color(t.prefixColor).content(prefix))
            scoreboardTeam.suffix(Component.empty())
            scoreboardTeam.setCanSeeFriendlyInvisibles(t.seeInvisible)
            scoreboardTeam.setAllowFriendlyFire(t.friendlyFire)
            scoreboardTeam.color(t.color)

            // Update all the players on the team
            val playersOnTeam = t.players.mapNotNull { Bukkit.getPlayer(it) }.map { it.name }
            val playersOnScoreboardTeam = scoreboardTeam.entries

            val toAdd = playersOnTeam.filter { p -> !playersOnScoreboardTeam.contains(p) }
            val toRemove = playersOnScoreboardTeam.filter { p -> !playersOnTeam.contains(p) }
            if (debug && toAdd.isNotEmpty())
                println("[ADD] " + toAdd.toTypedArray().contentToString())
            if (debug && toRemove.isNotEmpty())
                println("[REMOVE] " + toRemove.toTypedArray().contentToString())

            toRemove.forEach { scoreboardTeam.removeEntry(it) }
            toAdd.forEach { scoreboardTeam.addEntry(it) }
        }
        // Remove teams that no longer exist
        board.teams.forEach { team ->
            val exists = getTeams().any { it.teamName.startsWith(team.name) }
            if (!exists) {
                if (debug)
                    println("Unregistering team " + team.name)
                team.unregister()
            }
        }
    }

    /**
     * Adds an objective to the scoreboard
     *
     * @param displayName The display name of the objective
     * @param displaySlot The slot that the objective is displayed in
     * @param criteria    The criteria of the objective
     *
     * @return The [Objective] that was added
     */
    @JvmOverloads
    fun addObjective(
        displayName: Component,
        displaySlot: DisplaySlot,
        criteria: Criteria = Criteria.DUMMY
    ): Objective {
        val o = board.registerNewObjective("Obj-" + r.nextInt(999999), criteria, displayName)
        o.displaySlot = displaySlot
        return o
    }

    /**
     * Called after the scoreboard has been rendered
     */
    open fun postDraw() {

    }

    /**
     * Called before the scoreboard is rendered
     */
    open fun preDraw() {

    }

}
